src/components/cylc/Graph.vue

```diff
@@ -0,0 +1,1284 @@
+    this.handleMounted()
+    this.$store.watch((store) => {
+      this.workflows = store.workflows
+    }
+    )
+  },
+
+  created (cy) {
+    console.log('CREATED')
+    // |- websocket functionality ->
+    // this.$options.sockets.onmessage = (msg) => this.messageReceived(msg)
+    // --|
+    this.workflowId = this.$route.params.workflowid
+    workflowService.register(
+      this,
+      {
+        activeCallback: this.setActive
+      }
+    )
+    this.subscribe('root')
+  },
+
+  components: {
+    SyncLoader,
+    cytoscape: VueCytoscape
+  },
+
+  methods: {
+    subscribe (queryName) {
+      const id = workflowService.subscribe(
+        this,
+        QUERIES[queryName].replace('WORKFLOW_ID', this.workflowId),
+        this.setActive
+      )
+      if (!(queryName in this.subscriptions)) {
+        this.subscriptions[queryName] = {
+          id,
+          active: false
+        }
+      }
+    },
+
+    unsubscribe (queryName) {
+      if (queryName in this.subscriptions) {
+        workflowService.unsubscribe(
+          this.subscriptions[queryName].id
+        )
+      }
+    },
+
+    setActive (isActive) {
+      this.isLoading = !isActive
+    },
+
+    async workflowUpdated (workflows) {
+      try {
+        const elements = {
+          nodes: [],
+          edges: []
+        }
+        if (!isUndefined(workflows)) {
+          each(workflows, (value, key) => {
+            each(value, (workflow, key) => {
+              if (has(workflow.nodesEdges, 'edges') && !isUndefined(workflow.nodesEdges.edges)) {
+                elements.edges = this.getEdges(workflow.nodesEdges.edges)
+              }
+              if (has(workflow.nodesEdges, 'nodes') && !isUndefined(workflow.nodesEdges.nodes)) {
+                elements.nodes = this.getNodes(workflow.nodesEdges.nodes)
+              }
+            })
+          })
+        }
+        // console.log('elements ==>>>> ', elements)
+        isEmpty(elements) ? console.log('gdata is empty or undefined') : this.graphData = elements
+      } catch (error) {
+        console.error('workflowUpdated error: ', error)
+      }
+    },
+
+    getNodes (nodes) {
+      try {
+        const nodesArray = []
+        let nodeObj = {}
+        let parentId = ''
+        let todo = 0
+        each(nodes, (node, key) => {
+          nodeObj = {
+            data: {
+              id: '',
+              state: 'undefined',
+              parent: '',
+              label: '',
+              shape: 'ellipse',
+              runpercent: 0,
+              todo: 0
+            },
+            position: {
+
+            },
+            group: 'nodes',
+            removed: false,
+            selected: false,
+            selectable: true,
+            locked: false,
+            grabbable: true,
+            classes: ''
+          }
+          has(node, 'id') && !isEmpty(node.id) ? nodeObj.data.id = getUuid(node.id) : console.log('workflowUpdated - node id is empty')
```

Comment:
Can't we just use `node.id` when creating `nodeObj`? @dwsutherland can correct me if I am wrong, but in the whole `nodes`, each node will have an unique ID.

e.g. Workflow ID `five`... the cycle point ID 'five/20100101T000000Z`, the task proxy `five/20100101T000000Z/foo`, and the job `five/20100101T000000Z/foo/01`. They should not repeat within the same workflow? If so, we don't need to calculate the UUID I think.

For the Tree view, the `v-for` loop demands a `key`, that must be unique. I am using the `.id` of each entry in the workflows GraphQL response.

ID's are also required in our schema (quite sure somewhere in GraphiQL it should appear with a `!` ? See this part of the Cylc Flow code: https://github.com/cylc/cylc-flow/blob/f9927a5c53f072e24f65bd0f2355bff393af5b42/cylc/flow/network/schema.py#L419

Reply:
I found it easier to scan for different nodes with unique ids while developing, but I can keep those just for the edges if required

Reply:
Gotcha, it was more in case it helped with performance by avoiding computing the UUIDs, but I believe this is probably nanoseconds and doesn't really matter :+1:

Reply:
Every little helps :-)

Reply:
> Can't we just use node.id when creating nodeObj? @dwsutherland can correct me if I am wrong, but in the whole nodes, each node will have an unique ID.

Yes all nodes and edges have unique IDs  (i.e. `sutherlander|noz|20190101T00|baa`), if they have the same ID then they are duplicate elements (circular query, or multiple queries) or reference IDs to the respective data element.